File: packages/types/src/learning.ts
/**
 * Learning + Governance primitives — Phase 5 surface.
 *
 * The learning loop closes the circuit:
 *
 *   Setup + Fill ──► LearningEvent ──► ConfidenceCalibration
 *                          │
 *                          ├──► RegimeSnapshot (per symbol, per tf)
 *                          │
 *                          ├──► SessionSnapshot (london/ny/asia)
 *                          │
 *                          └──► StrategyDNA (where it works / fails)
 *
 * A separate DataTruthCheck monitor watches the ingestion fabric for
 * bad feeds, latency, gaps, and trips the kill switch when any gate
 * fails. The learning agent only promotes strategies when data truth
 * is green.
 *
 * Wire-shape notes:
 *   * Every timestamp is an ISO-8601 UTC string (Z suffix).
 *   * camelCase over the wire — Pydantic v2 models use populate_by_name.
 *   * Calibration curves and DNA grids are dense numeric arrays; no
 *     sparse envelopes (the dimensions are small + fixed).
 */
import { z } from "zod";
import { TimeframeSchema } from "./market.js";
import { SetupTypeSchema } from "./setups.js";
import { StrategyTierSchema } from "./quant-lab.js";

// ──────────────────────────── learning events ──────────────────────────

/**
 * The canonical event types emitted to the learning bus. One row per
 * event. Events are immutable and the learning worker consumes them
 * in strict createdAt order.
 */
export const LearningEventKindSchema = z.enum([
  "setup_detected",
  "setup_approved",
  "setup_rejected",
  "trade_opened",
  "trade_closed_win",
  "trade_closed_loss",
  "trade_closed_scratch",
  "backtest_completed",
  "calibration_updated",
  "regime_flipped",
  "data_truth_breach",
  "promotion_auto_demote",
  "promotion_manual",
]);
export type LearningEventKind = z.infer<typeof LearningEventKindSchema>;

export const LearningEventSchema = z.object({
  id: z.string().min(1),
  kind: LearningEventKindSchema,
  /** Subject id — setupId | tradeId | backtestId | strategyId. */
  subjectId: z.string().min(1),
  subjectKind: z.enum([
    "setup",
    "paper_trade",
    "live_trade",
    "backtest",
    "strategy",
    "calibration",
    "regime",
    "data_truth",
  ]),
  strategyId: z.string().nullable(),
  /** Arbitrary JSON payload — kind-specific. */
  payload: z.record(z.string(), z.any()).default({}),
  /** Optional correlation id grouping related events. */
  correlationId: z.string().nullable(),
  occurredAt: z.string().datetime(),
  ingestedAt: z.string().datetime(),
});
export type LearningEvent = z.infer<typeof LearningEventSchema>;

export const LearningEventFilterSchema = z.object({
  kind: LearningEventKindSchema.optional(),
  subjectKind: LearningEventSchema.shape.subjectKind.optional(),
  strategyId: z.string().optional(),
  fromTs: z.string().datetime().optional(),
  toTs: z.string().datetime().optional(),
  cursor: z.string().datetime().optional(),
  limit: z.number().int().min(1).max(500).default(100),
});
export type LearningEventFilter = z.infer<typeof LearningEventFilterSchema>;

export const LearningEventsListOutSchema = z.object({
  events: z.array(LearningEventSchema),
  total: z.number().int().nonnegative(),
});
export type LearningEventsListOut = z.infer<typeof LearningEventsListOutSchema>;

// ──────────────────────────── confidence calibration ───────────────────

/**
 * Calibrator kind. `bucket` is the bit-stable default (isotonic buckets
 * over 10 bins of raw score). `platt` is a 2-parameter sigmoid fit for
 * strategies with enough samples (>= 200).
 */
export const CalibrationKindSchema = z.enum(["bucket", "platt"]);
export type CalibrationKind = z.infer<typeof CalibrationKindSchema>;

/**
 * One bin on a bucket calibrator. `rawLow..rawHigh` maps to a calibrated
 * probability `calibrated` with sample size `count`.
 */
export const CalibrationBinSchema = z.object({
  rawLow: z.number().min(0).max(1),
  rawHigh: z.number().min(0).max(1),
  calibrated: z.number().min(0).max(1),
  count: z.number().int().nonnegative(),
  wins: z.number().int().nonnegative(),
});
export type CalibrationBin = z.infer<typeof CalibrationBinSchema>;

/**
 * Calibrator snapshot for a given (strategyId, setupType, tf) key.
 * Recomputed hourly; the most recent row is canonical.
 */
export const ConfidenceCalibrationSchema = z.object({
  id: z.string().min(1),
  strategyId: z.string().nullable(),
  setupType: SetupTypeSchema.nullable(),
  tf: TimeframeSchema.nullable(),
  kind: CalibrationKindSchema,
  /** Bucket kind only — 10 bins from 0..1. */
  bins: z.array(CalibrationBinSchema),
  /** Platt kind only — calibrated = 1 / (1 + exp(a*raw + b)). */
  plattA: z.number().nullable(),
  plattB: z.number().nullable(),
  /** Expected Calibration Error — lower is better. */
  ece: z.number().min(0).max(1),
  /** Brier score — lower is better. */
  brier: z.number().min(0).max(1),
  /** Sample size that fit this curve. */
  sampleSize: z.number().int().nonnegative(),
  generatedAt: z.string().datetime(),
});
export type ConfidenceCalibration = z.infer<typeof ConfidenceCalibrationSchema>;

export const CalibrationCurvesOutSchema = z.object({
  curves: z.array(ConfidenceCalibrationSchema),
  generatedAt: z.string().datetime(),
});
export type CalibrationCurvesOut = z.infer<typeof CalibrationCurvesOutSchema>;

// ──────────────────────────── regime detection ─────────────────────────

export const RegimeKindSchema = z.enum([
  "trending",
  "ranging",
  "volatile",
  "news_driven",
]);
export type RegimeKind = z.infer<typeof RegimeKindSchema>;

/**
 * Per (symbolId, tf) regime verdict. Emitted by the regime detector on
 * the structure + order-flow stream. Updated on every fresh bar close;
 * canonical row is the latest per (symbolId, tf).
 */
export const RegimeSnapshotSchema = z.object({
  id: z.string().min(1),
  symbolId: z.string().min(1),
  tf: TimeframeSchema,
  kind: RegimeKindSchema,
  /** 0..1 — how strongly the detector is asserting this regime. */
  confidence: z.number().min(0).max(1),
  /** The signed trend strength (-1..1). */
  trendStrength: z.number().min(-1).max(1),
  /** 0..1 normalised volatility percentile. */
  volatility: z.number().min(0).max(1),
  /** Wall-clock age of the bar that produced this verdict. */
  barAgeMs: z.number().int().nonnegative(),
  observedAt: z.string().datetime(),
  /** Optional short narrative. */
  notes: z.string().max(2000).default(""),
});
export type RegimeSnapshot = z.infer<typeof RegimeSnapshotSchema>;

export const RegimeCurrentOutSchema = z.object({
  snapshots: z.array(RegimeSnapshotSchema),
  generatedAt: z.string().datetime(),
});
export type RegimeCurrentOut = z.infer<typeof RegimeCurrentOutSchema>;

export const RegimeHistoryFilterSchema = z.object({
  symbolId: z.string().min(1),
  tf: TimeframeSchema,
  fromTs: z.string().datetime().optional(),
  toTs: z.string().datetime().optional(),
  limit: z.number().int().min(1).max(500).default(200),
});
export type RegimeHistoryFilter = z.infer<typeof RegimeHistoryFilterSchema>;

export const RegimeHistoryOutSchema = z.object({
  symbolId: z.string().min(1),
  tf: TimeframeSchema,
  snapshots: z.array(RegimeSnapshotSchema),
});
export type RegimeHistoryOut = z.infer<typeof RegimeHistoryOutSchema>;

// ──────────────────────────── session intelligence ─────────────────────

export const TradingSessionSchema = z.enum([
  "asia",
  "london",
  "ny_am",
  "ny_pm",
  "off_hours",
]);
export type TradingSession = z.infer<typeof TradingSessionSchema>;

export const SessionSnapshotSchema = z.object({
  id: z.string().min(1),
  symbolId: z.string().min(1),
  session: TradingSessionSchema,
  /** Mean realised volatility (normalised 0..1) in this session. */
  volatility: z.number().min(0).max(1),
  /** Historical win rate of our setups played in this session. */
  winRate: z.number().min(0).max(1).nullable(),
  /** Mean R across our setups played in this session. */
  meanR: z.number().nullable(),
  sampleSize: z.number().int().nonnegative(),
  observedAt: z.string().datetime(),
});
export type SessionSnapshot = z.infer<typeof SessionSnapshotSchema>;

export const SessionIntelOutSchema = z.object({
  snapshots: z.array(SessionSnapshotSchema),
  generatedAt: z.string().datetime(),
});
export type SessionIntelOut = z.infer<typeof SessionIntelOutSchema>;

// ──────────────────────────── data truth + kill switch ─────────────────

export const DataTruthStatusSchema = z.enum(["green", "amber", "red"]);
export type DataTruthStatus = z.infer<typeof DataTruthStatusSchema>;

export const DataTruthCheckKindSchema = z.enum([
  "bar_latency",
  "bar_gap",
  "book_staleness",
  "feed_desync",
  "symbol_missing",
  "broker_heartbeat",
]);
export type DataTruthCheckKind = z.infer<typeof DataTruthCheckKindSchema>;

export const DataTruthCheckSchema = z.object({
  id: z.string().min(1),
  kind: DataTruthCheckKindSchema,
  status: DataTruthStatusSchema,
  /** Short narrative — what's failing + what it implies. */
  message: z.string().max(2000),
  /** Numeric measurement — latency ms, gap s, etc. */
  measurement: z.number(),
  /** Configured threshold for amber. */
  amberThreshold: z.number(),
  /** Configured threshold for red (kill-switch trip). */
  redThreshold: z.number(),
  symbolId: z.string().nullable(),
  observedAt: z.string().datetime(),
});
export type DataTruthCheck = z.infer<typeof DataTruthCheckSchema>;

export const DataTruthStatusOutSchema = z.object({
  /** Overall status = worst of any individual check. */
  status: DataTruthStatusSchema,
  checks: z.array(DataTruthCheckSchema),
  /** If true, live-gate is unconditionally rejecting new approvals. */
  killSwitchTripped: z.boolean(),
  /** Short reason the kill-switch is tripped (null if not tripped). */
  killSwitchReason: z.string().nullable(),
  generatedAt: z.string().datetime(),
});
export type DataTruthStatusOut = z.infer<typeof DataTruthStatusOutSchema>;

// ──────────────────────────── strategy DNA ─────────────────────────────

/**
 * Compact 2-D grid summarising where a strategy has worked historically.
 * Rows = regime, columns = session. Each cell carries win rate + mean R
 * + sample size. Updated as trades accumulate.
 */
export const DNACellSchema = z.object({
  regime: RegimeKindSchema,
  session: TradingSessionSchema,
  winRate: z.number().min(0).max(1).nullable(),
  meanR: z.number().nullable(),
  sampleSize: z.number().int().nonnegative(),
});
export type DNACell = z.infer<typeof DNACellSchema>;

export const StrategyDNASchema = z.object({
  id: z.string().min(1),
  strategyId: z.string().min(1),
  cells: z.array(DNACellSchema),
  /** Rollup of best + worst cells for quick rendering. */
  bestCell: DNACellSchema.nullable(),
  worstCell: DNACellSchema.nullable(),
  /** Tier snapshot at generation time (not authoritative — see Strategy). */
  tierAtGeneration: StrategyTierSchema,
  totalTrades: z.number().int().nonnegative(),
  generatedAt: z.string().datetime(),
});
export type StrategyDNA = z.infer<typeof StrategyDNASchema>;

export const StrategyDNAListOutSchema = z.object({
  dna: z.array(StrategyDNASchema),
  generatedAt: z.string().datetime(),
});
export type StrategyDNAListOut = z.infer<typeof StrategyDNAListOutSchema>;

File: services/control_plane/app/routes/learning.py
"""Learning + governance HTTP surface — Phase 5 PR8.

Wire contract:

  * ``GET  /v1/learning/events``                 — append-only event log.
  * ``GET  /v1/learning/calibration``            — latest curves per scope.
  * ``POST /v1/learning/calibration/recompute``  — admin recompute trigger.
  * ``GET  /v1/learning/regime``                 — current regime verdicts.
  * ``GET  /v1/learning/regime/history``         — history for symbol+tf.
  * ``GET  /v1/learning/sessions``               — per-session rollups.
  * ``GET  /v1/learning/data-truth``             — health + kill-switch.
  * ``POST /v1/learning/data-truth/checks``      — admin write a check.
  * ``GET  /v1/learning/dna``                    — strategy DNA grids.
  * ``POST /v1/learning/dna/{strategyId}/rebuild``— admin recompute DNA.

Reads accept either snake_case or camelCase parameters. Writes go
through :func:`app.audit.log_event` with the relevant
``resource_type``. Reads are un-logged to keep the audit stream
readable.

Error codes:
  * ``invalid_event_kind``        — unknown kind on GET /events.
  * ``invalid_subject_kind``      — unknown subject on GET /events.
  * ``invalid_data_truth_kind``   — unknown check kind on POST checks.
  * ``invalid_threshold_pair``    — amber >= red on POST checks.
  * ``strategy_not_found``        — POST /dna/{id}/rebuild, id missing.
  * ``regime_query_missing``      — GET /regime/history, symbolId+tf required.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter, Query, Request, status

from app.audit import log_event
from app.db import DbSession
from app.deps import AdminUser, CurrentUser
from app.errors import ApiError
from app.learning.dto import (
    CalibrationCurvesOut,
    CalibrationRecomputeRequestDto,
    CalibrationRecomputeResponseDto,
    DataTruthCheckCreateRequestDto,
    DataTruthCheckDto,
    DataTruthStatusOutDto,
    LearningEventsListOut,
    RegimeCurrentOutDto,
    RegimeHistoryOutDto,
    SessionIntelOutDto,
    StrategyDNADto,
    StrategyDNAListOutDto,
)
from app.learning.repo import (
    build_dna_for_strategy,
    get_data_truth_status,
    list_learning_events,
    list_regime_history,
    list_regime_snapshots,
    list_session_snapshots,
    list_strategy_dna,
    load_calibration_curves,
    recompute_calibration,
    upsert_data_truth_check,
)


UTC = timezone.utc

router = APIRouter(prefix="/learning", tags=["learning"])


_VALID_EVENT_KINDS = frozenset(
    {
        "setup_detected",
        "setup_approved",
        "setup_rejected",
        "trade_opened",
        "trade_closed_win",
        "trade_closed_loss",
        "trade_closed_scratch",
        "backtest_completed",
        "calibration_updated",
        "regime_flipped",
        "data_truth_breach",
        "promotion_auto_demote",
        "promotion_manual",
    }
)
_VALID_SUBJECT_KINDS = frozenset(
    {
        "setup",
        "paper_trade",
        "live_trade",
        "backtest",
        "strategy",
        "calibration",
        "regime",
        "data_truth",
    }
)
_VALID_DATA_TRUTH_KINDS = frozenset(
    {
        "bar_latency",
        "bar_gap",
        "book_staleness",
        "feed_desync",
        "symbol_missing",
        "broker_heartbeat",
    }
)
_VALID_SESSIONS = frozenset(
    {"asia", "london", "ny_am", "ny_pm", "off_hours"}
)


def _now_utc() -> datetime:
    return datetime.now(UTC)


# ─────────────────────────── learning events ───────────────────────────


@router.get(
    "/events",
    response_model=LearningEventsListOut,
    summary="List learning events (newest first)",
)
async def list_events(
    db: DbSession,
    user: CurrentUser,
    kind: str | None = Query(default=None),
    subject_kind: str | None = Query(default=None, alias="subjectKind"),
    strategy_id: str | None = Query(default=None, alias="strategyId"),
    from_ts: datetime | None = Query(default=None, alias="fromTs"),
    to_ts: datetime | None = Query(default=None, alias="toTs"),
    cursor: datetime | None = Query(default=None),
    limit: int = Query(default=100, ge=1, le=500),
) -> LearningEventsListOut:
    if kind is not None and kind not in _VALID_EVENT_KINDS:
        raise ApiError(
            status_code=status.HTTP_400_BAD_REQUEST,
            code="invalid_event_kind",
            message=f"kind must be one of {sorted(_VALID_EVENT_KINDS)}",
        )
    if subject_kind is not None and subject_kind not in _VALID_SUBJECT_KINDS:
        raise ApiError(
            status_code=status.HTTP_400_BAD_REQUEST,
            code="invalid_subject_kind",
            message=f"subjectKind must be one of {sorted(_VALID_SUBJECT_KINDS)}",
        )
    return await list_learning_events(
        db,
        kind=kind,  # type: ignore[arg-type]
        subject_kind=subject_kind,  # type: ignore[arg-type]
        strategy_id=strategy_id,
        from_ts=from_ts,
        to_ts=to_ts,
        cursor=cursor,
        limit=limit,
    )


# ─────────────────────────── calibration ───────────────────────────────


@router.get(
    "/calibration",
    response_model=CalibrationCurvesOut,
    summary="List latest calibration curves per scope",
)
async def get_calibration(
    db: DbSession,
    user: CurrentUser,
) -> CalibrationCurvesOut:
    return await load_calibration_curves(db)


@router.post(
    "/calibration/recompute",
    response_model=CalibrationRecomputeResponseDto,
    status_code=status.HTTP_201_CREATED,
    summary="Recompute the calibration curve for a scope (admin only)",
)
async def recompute_calibration_route(
    request: Request,
    req: CalibrationRecomputeRequestDto,
    db: DbSession,
    admin: AdminUser,
) -> CalibrationRecomputeResponseDto:
    response = await recompute_calibration(db, req=req)
    await log_event(
        db,
        request=request,
        actor_user_id=admin.id,
        actor_email=admin.email,
        action="learning.calibration.recompute",
        resource_type="learning.calibration",
        resource_id=response.curve.id,
        outcome="success",
        details={
            "strategyId": req.strategy_id,
            "setupType": req.setup_type,
            "tf": req.tf,
            "kind": req.kind,
            "sampleSize": response.sample_size,
        },
    )
    await db.commit()
    return response


# ─────────────────────────── regime ────────────────────────────────────


@router.get(
    "/regime",
    response_model=RegimeCurrentOutDto,
    summary="List current regime verdicts (latest per symbolId + tf)",
)
async def get_regime(
    db: DbSession,
    user: CurrentUser,
) -> RegimeCurrentOutDto:
    return await list_regime_snapshots(db)


@router.get(
    "/regime/history",
    response_model=RegimeHistoryOutDto,
    summary="Regime history for one (symbolId, tf)",
)
async def get_regime_history(
    db: DbSession,
    user: CurrentUser,
    symbol_id: str | None = Query(default=None, alias="symbolId"),
    tf: str | None = Query(default=None),
    from_ts: datetime | None = Query(default=None, alias="fromTs"),
    to_ts: datetime | None = Query(default=None, alias="toTs"),
    limit: int = Query(default=200, ge=1, le=500),
) -> RegimeHistoryOutDto:
    if not symbol_id or not tf:
        raise ApiError(
            status_code=status.HTTP_400_BAD_REQUEST,
            code="regime_query_missing",
            message="symbolId and tf are required",
        )
    return await list_regime_history(
        db,
        symbol_id=symbol_id,
        tf=tf,
        from_ts=from_ts,
        to_ts=to_ts,
        limit=limit,
    )


# ─────────────────────────── sessions ──────────────────────────────────


@router.get(
    "/sessions",
    response_model=SessionIntelOutDto,
    summary="Per-session snapshots",
)
async def get_sessions(
    db: DbSession,
    user: CurrentUser,
    symbol_id: str | None = Query(default=None, alias="symbolId"),
    session: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
) -> SessionIntelOutDto:
    if session is not None and session not in _VALID_SESSIONS:
        raise ApiError(
            status_code=status.HTTP_400_BAD_REQUEST,
            code="invalid_session",
            message=f"session must be one of {sorted(_VALID_SESSIONS)}",
        )
    return await list_session_snapshots(
        db, symbol_id=symbol_id, session=session, limit=limit
    )


# ─────────────────────────── data truth ────────────────────────────────


@router.get(
    "/data-truth",
    response_model=DataTruthStatusOutDto,
    summary="Data-truth status + kill-switch",
)
async def get_data_truth(
    db: DbSession,
    user: CurrentUser,
) -> DataTruthStatusOutDto:
    return await get_data_truth_status(db)


@router.post(
    "/data-truth/checks",
    response_model=DataTruthCheckDto,
    status_code=status.HTTP_201_CREATED,
    summary="Write a data-truth check (admin only)",
)
async def post_data_truth_check(
    request: Request,
    req: DataTruthCheckCreateRequestDto,
    db: DbSession,
    admin: AdminUser,
) -> DataTruthCheckDto:
    if req.kind not in _VALID_DATA_TRUTH_KINDS:
        raise ApiError(
            status_code=status.HTTP_400_BAD_REQUEST,
            code="invalid_data_truth_kind",
            message=f"kind must be one of {sorted(_VALID_DATA_TRUTH_KINDS)}",
        )
    if req.amber_threshold > req.red_threshold:
        raise ApiError(
            status_code=status.HTTP_400_BAD_REQUEST,
            code="invalid_threshold_pair",
            message="amberThreshold must be ≤ redThreshold",
        )

    row, check_status = await upsert_data_truth_check(db, req=req)
    await log_event(
        db,
        request=request,
        actor_user_id=admin.id,
        actor_email=admin.email,
        action="learning.data_truth.write",
        resource_type="learning.data_truth",
        resource_id=row.id,
        outcome="success",
        details={
            "kind": req.kind,
            "measurement": req.measurement,
            "amberThreshold": req.amber_threshold,
            "redThreshold": req.red_threshold,
            "status": check_status,
        },
    )
    await db.commit()
    return DataTruthCheckDto(
        id=row.id,
        kind=req.kind,
        status=check_status,
        message=req.message or "",
        measurement=req.measurement,
        amberThreshold=req.amber_threshold,
        redThreshold=req.red_threshold,
        symbolId=req.symbol_id,
        observedAt=row.last_observed_at,
    )


# ─────────────────────────── strategy DNA ──────────────────────────────


@router.get(
    "/dna",
    response_model=StrategyDNAListOutDto,
    summary="List strategy DNA grids",
)
async def get_dna(
    db: DbSession,
    user: CurrentUser,
    strategy_id: str | None = Query(default=None, alias="strategyId"),
) -> StrategyDNAListOutDto:
    return await list_strategy_dna(db, strategy_id=strategy_id)


@router.post(
    "/dna/{strategy_id}/rebuild",
    response_model=StrategyDNADto,
    status_code=status.HTTP_201_CREATED,
    summary="Rebuild the DNA grid for one strategy (admin only)",
)
async def rebuild_dna(
    strategy_id: str,
    request: Request,
    db: DbSession,
    admin: AdminUser,
) -> StrategyDNADto:
    try:
        dto = await build_dna_for_strategy(db, strategy_id=strategy_id)
    except LookupError:
        raise ApiError(
            status_code=status.HTTP_404_NOT_FOUND,
            code="strategy_not_found",
            message=f"no strategy with id {strategy_id!r}",
        )
    await log_event(
        db,
        request=request,
        actor_user_id=admin.id,
        actor_email=admin.email,
        action="learning.dna.rebuild",
        resource_type="learning.dna",
        resource_id=strategy_id,
        outcome="success",
        details={
            "strategyId": strategy_id,
            "totalTrades": dto.total_trades,
        },
    )
    await db.commit()
    return dto
